db/compaction/compaction_job.cc

```diff
@@ -1575,15 +1079,17 @@ void CompactionJob::ProcessKeyValueCompaction(SubcompactionState* sub_compact) {
     if (c_iter->status().IsManualCompactionPaused()) {
       break;
     }
+
+    // TODO: better handle preclude_bottommost output file cut
     if (!output_file_ended && c_iter->Valid()) {
       if (((partitioner.get() &&
             partitioner->ShouldPartition(PartitionerRequest(
                 last_key_for_partitioner, c_iter->user_key(),
-                sub_compact->current_output_file_size)) == kRequired) ||
+                sub_compact->GetCurrentOutputFileSize())) == kRequired) ||
            (sub_compact->compaction->output_level() != 0 &&
             sub_compact->ShouldStopBefore(
-                c_iter->key(), sub_compact->current_output_file_size))) &&
-          sub_compact->builder != nullptr) {
+                c_iter->key(), sub_compact->GetCurrentOutputFileSize()))) &&
+          sub_compact->HasBuilder()) {
```

Comment:
I feel that it might be easier to read if the logic of managing files in each output level (bottommost/precluded) is encapsulated in a separate class. Now the logic of handling file cutting, file creating, file closing + checksuming, etc, is tangled with key processing and is hard to reason about. Separating it out will make it easier for me to review the change. Knowing GetTableBuilder() and GetFileWriter() can mean file from either output level makes it hard to read